File: djsp.c
#include <stdio.h>
#include <getopt.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include "include/defs.h"
#include "include/logging.h"
#include "include/seq.h"



unsigned long parse_int(
  char *str
) {
  errno = 0;
  char *end;
  unsigned long ret = strtoul(str, &end, 0);

  if (errno)
    DIE("unable to convert '%s' to an integer: %s\n", str, strerror(errno));

  if (*end != '\0') {
    if (end == str)
      DIE("unable to find an integer in '%s'\n", str);
    else
      fprintf(stderr, "trailing junk '%s' in '%s'\n", end, str);
  }

  if (end == str)
    DIE("empty integer given\n");

  return ret;
}

FILE *open_file(
  char *path
) {
  FILE *out = fopen(path, "a");
  if (out == NULL)
    DIE("unable to open file '%s': %s\n", path, STRERRNO());

  setvbuf(out, NULL, _IOLBF, 0);

  return out;
}




static const char *usage =
  "usage: %s [short options]\n"
  "-l --log                    [f = filename]    set the output log file to 'f'\n"
  "-t --time-format            [t = time format] set the output time format to 't'\n"
  "-s --steps                  [n = integer]     show intermediate steps for values which exceed 2^'n'\n"
  "-L --start                  [n = integer]     start running from 'n'\n"
  "-R --end                    [n = integer]     stop running after 'n'\n"
  "-1 --oneshot (only with -L)                   only run for one integer\n"
  "-u --before-up                                run bignum reallocations before the up step\n"
  "-U --after-up                                 run bignum reallocations after the up step\n"
  "-d --before-down                              run bignum reallocations before the down step\n"
  "-D --after-down                               run bignum reallocations after the down step\n"
;

// investigate "suckless getopt":
// https://lobste.rs/s/pd7zaw/arg_h_suckless_alternative_getopt

int main(
  int argc,
  char **argv
) {
  opterr = 0;

  static struct option long_options[] = {
    { "log",         required_argument, NULL,                        'l'  },
    { "time-format", required_argument, NULL,                        't'  },
    { "steps",       optional_argument, NULL,                        's'  },
    { "start",       required_argument, NULL,                        'L'  },
    { "end",         required_argument, NULL,                        'R'  },
    { "oneshot",     no_argument,       NULL,                        '1'  },
    { "before-up",   no_argument,       (int *)&realloc_before_up,   true },
    { "after-up",    no_argument,       (int *)&realloc_after_up,    true },
    { "before-down", no_argument,       (int *)&realloc_before_down, true },
    { "after-down",  no_argument,       (int *)&realloc_after_down,  true },
    { 0, 0, 0, 0 }
  };

  for (int option_index, c; (c = getopt_long(
    argc, argv,
    "l:" "t:" "s::" "L:R:1" "dDuU",
    long_options, &option_index
  )) != -1; ) {
    switch (c) {
      case 0: {
        if (long_options[option_index].flag)
          break;

        fprintf(stderr, "hit 0!?\n");
        abort();
      } break;

      case 'l': log_file = open_file(optarg); break;

      case 't': time_format = optarg; break;

      case 's': {
        if (optarg)
          ssol = parse_int(optarg);
        else if (argv[optind] && argv[optind][0] != '-')
          ssol = parse_int(argv[optind++]);
        else
          ssol = 0;

        show_steps = 1;
      } break;

      case 'L': start_value = parse_int(optarg); break;
      case 'R': {
        end_value = parse_int(optarg);
        ending = 1;
      } break;
      case '1': {
        if (start_value == 1) {
          fprintf(stderr, "oneshot doesn't make sense without a starting value\n\n");
          fprintf(stderr, usage, *argv);
          exit(1);
        }

        end_value = start_value;
        ending = 1;
      } break;

      case 'd': realloc_before_down = true; break;
      case 'D': realloc_after_down  = true; break;
      case 'u': realloc_before_up   = true; break;
      case 'U': realloc_after_up    = true; break;

      case '?': {
        fprintf(stderr, "unrecognized option '%s'\n\n", argv[optind - 1]);
        fprintf(stderr, usage, *argv);
        exit(1);
      } break;

      default: DIE("oh nooo! default hit!\n");
    }
  }

  if (optind < argc)
    fprintf(stderr, "ignoring trailing arguments...\n");

  // now... sanity check!

  message("hello, world!\n");

  if (!show_steps) {
    printf("NOT showing steps!");
  } else {
    printf("showing steps");
    if (ssol)
      printf(" above 2^%lu", ssol);
  }
  printf("\n");

  printf("running on %lu", start_value);
  if (ending) {
    if (end_value != start_value)
      printf(" -> %lu", end_value);
  } else
    printf(" -> oo");
  printf("\n");

  printf(
    "%cO%c %cE%c\n",
    realloc_before_up ? 'Y' : 'N', realloc_after_up ? 'Y' : 'N',
    realloc_before_down ? 'Y' : 'N', realloc_after_down ? 'Y' : 'N'
  );
}

File: djsp.rb
#!/usr/bin/env ruby

require "ffi"

module DJSP
  module C
    extend FFI::Library
    ffi_lib "c"
    ffi_lib "./build/libdjsp.so"

    # step

    [
      :step_realloc_before_up,   :step_realloc_after_up,
      :step_realloc_before_down, :step_realloc_after_down
    ].each {|name| attach_variable name, :bool }

    # logging
    
    attach_function :setup, [], :void
    attach_function :set_log, [:string], :void
    attach_function :djsp_message, [:string, :varargs], :void

    # oneshot

    attach_variable :show_steps, :bool
    attach_variable :ssol, :size_t

    # callback :memo_callback, [:uint64, :uint64, :pointer], :uint64
    attach_function :oneshot_2, [:uint64], :uint64
    attach_function :oneshot_n, [:uint64, :uint64], :uint64

    # sequence

		callback :memo_read_callback, [:uint64], :uint64
		callback :memo_write_callback, [*[:uint64] * 2], :void
    callback :hwm_callback, [*[:uint64] * 3], :void
    attach_function :sequence,           [:uint64, :uint64, :uint64, :memo_read_callback, :memo_write_callback, :hwm_callback], :void
    attach_function :sequence_2_rootopt, [         :uint64, :uint64, :memo_read_callback, :memo_write_callback, :hwm_callback], :void
    attach_function :sequence_2_p,       [         :uint64, :uint64, :memo_read_callback, :memo_write_callback, :hwm_callback], :void
  end

  public

  VERSION_MAJOR = 1
  VERSION_MINOR = 0
  VERSION_PATCH = 0
  VERSION       = "1.0.0"

  VALID_SEQUENCE_OPTIMIZATIONS = [:root, :none]

  class << self
    def ssol; C::ssol; end
    def ssol= x
      if x.nil?
        C::ssol = 0
        C::show_steps = false
      end

      raise ArgumentError unless x.is_a? Integer

      C::ssol = x
      C::show_steps = true
    end
    alias show_steps_over_log ssol
    alias show_steps_over_log= ssol=

    @log = nil
    def log; @log; end
    def log= file
      # clear set log file
      if file.nil?
        @log = nil
        C::set_log nil
        return
      end

      path =
        case file
        when File then file.to_path
        when String then file
        else raise ArgumentError, "the log file must be a File or a path to open"
        end

      # open file, ensure that ruby doesn't buffer any writes
      # (try this with sync = true and sync = false):
      # file.write "i <3 sweet exorcist's testone"
      # `kill -9 $(pgrep ruby)`
      # file.write "and testtwo\n"
      @log = File.new(path, "a").tap { _1.sync = true }

      C::set_up_log path
    end

    def message what
      time = Time.now.strftime "%FT%T.%L%z"
      puts "#{time}: #{what}"
      @log&.puts "#{time}: #{what}"
    end

    def oneshot seed, base: nil, &memo
      case base
      when nil then C::oneshot_2 seed
      else          C::oneshot_n seed, base
      end
    end

    def sequence range = (1..), base: nil, optimize: nil, alert: nil, &hwm
      unless range.is_a? Enumerator::ArithmeticSequence
      	begin
      		range = range.step
      		raise unless range.is_a? Enumerator::ArithmeticSequence
      	rescue
        	raise TypeError, "`range` must be or be able to be converted to an Enumerator::ArithmeticSequence"
        end
      end

      left = range.begin || 1
      (right, endless) = range.end ? [range.end, false] : [0, true]
      step = range.step

      base ||= 0

			unless optimize.nil?
		    optimize = optimize.to_sym
		    raise ArgumentError, "the only valid optimizations are: #{VALID_SEQUENCE_OPTIMIZATIONS.join ", "}" \
		      unless VALID_SEQUENCE_OPTIMIZATIONS.include? optimize
			end

      if block_given?
      	hwm = FFI::Function.new :void, [*[:uint64] * 3], hwm
      else
        out = {}
        hwm = FFI::Function.new(:void, [*[:uint64] * 3]) do |index, mark, where|
        	STDERR.puts "ruby entering function (#{index}, #{mark}, #{where})"
        	out[where] = mark
        	STDERR.puts "ruby exiting function (#{index}, #{mark}, #{where})"
        end
      end

      if optimize == :root
        raise RangeError, "root optimization not implemented for bases other than 2" \
          if base > 2

				STDERR.puts "sequence_2_rootopt with [#{left}, #{right}, nil, nil, #{hwm}]"
        C::sequence_2_rootopt left, right, nil, nil, hwm
      elsif optimize == :none || base > 2
				STDERR.puts "sequence with [#{base}, #{left}, #{right}, nil, nil, #{hwm}]"
        C::sequence base, left, right, nil, nil, hwm
      else
				STDERR.puts "sequence_2_p with [#{left}, #{right}, nil, nil, #{hwm}]"
				C::sequence_2_p left, right, nil, nil, hwm
      end

      out unless block_given?
    end

    def equal_to(what, base: nil) = (1..).lazy.filter { oneshot(_1, base: base) == what } 
    def first_equal_to(what, base: nil) = equal_to(what, base: base).first
    def equal_to_over range, base: nil, all: false
      memo = {}

      range.each do |i|
        res = oneshot i, base: base
        
        if all
          memo[res] ||= []
          memo[res] << i
        else
          memo[res] ||= i
        end
      end

      memo.sort_by {|k, _| k }.to_h
    end
  end

  C::setup
  freeze
end

if $0 == __FILE__
  require "optparse"

  options = {}
  OptionParser.new do |parser|
    parser.banner = "usage: #$0 [options]"

    parser.on_head(
      "-h", "--help",
      "show this help"
    ) { puts parser; exit }

    parser.on(
      "-l LOG", "--log LOG",
      "tee all output to LOG"
    ) {|log| DJSP.log = log }

    parser.on(
      "-s [BL]", "--steps [BL]", Integer,
      "show intermediate steps for values which are greater than this (B)ase-2 (L)ogarithm"
    ) {|steps| DJSP.show_steps_over_log = steps || 0 }

    parser.on(
      "-b [BASE = 2]", "--base [BASE = 2]", Integer,
      "set the base used for calculations to BASE"
    ) {|base| options[:base] = base || 2 }

    parser.on(
      "-a [STEP = 1]", "--alert-every [STEP = 1]", Integer,
      "alert when the sequence has elapsed STEP values"
    ) {|alert| options[:alert] = alert || 1 }
    
    [
      [?u, "before", "up"  ],
      [?U, "after",  "up"  ],
      [?d, "before", "down"],
      [?D, "after",  "down"]
    ].each do |(flag, where, step)|
      parser.on(
        "-#{flag}", "--#{where}-#{step}", TrueClass,
        "run bignum reallocations #{where} the #{step} step"
      ) { DJSP::C.send :"step_realloc_#{where}_#{step}=", true }
    end
  end.parse!

  raise OptionParser::InvalidOption, ARGV[1] if ARGV.length > 1
  
  if ARGV.length == 1 
    options[:over] =
      begin
        Integer ARGV.first
      rescue ArgumentError
        ARGV.first =~ /(\d*)\.\.(\d*)(?:,(\d+))?/
        raise OptionParser::InvalidOption, ARGV.first unless $~

        (start, ending) = [$1, $2].map { _1.empty? ? nil : _1.to_i }
        range = (start..ending)
        range = range.step $3.to_i if $3
        range
      end
  end

  options[:over] ||= (1..)

  if options[:over].is_a? Enumerable
    DJSP.sequence options[:over], base: options[:base], alert: options[:alert] do |index, mark, where|
      DJSP.message "A#{options[:base] ? "_#{options[:base]}" : ""}(#{index}) @ #{where} = #{mark}"
    end
  else
    r = DJSP.oneshot options[:over], base: options[:base]
    DJSP.message "J#{options[:base] ? "_#{options[:base]}" : ""}(#{options[:over]}) = #{r}"
  end
end
